### ISSUE TITLE
Would like to experiment with modulation indication

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#136]
Is there a notification for modulation update that the graphics can use so that repaint can be called for given widget? Or maybe repaint is not called per-widget...

ok you have this:

![image](https://user-images.githubusercontent.com/9116575/33977334-9f6050ac-e057-11e7-82da-127def43abc1.png)

but that's only for modulation assignment, not modulation value. You should probably rename that function to "ModulationAssignmentChange" or "ConnecitonChange".

I am working on a knob widget and in its paint function I have this:

![image](https://user-images.githubusercontent.com/9116575/33977397-ef79d1ee-e057-11e7-8382-c7252b2b78ec.png)

so all I need to do is fill out that modulatedValue somehow, and the repaint needs to be called when modulation value updates.

**Comment by elanhickler** [comment]
heehee it's working!!!!!!!!... well, only when I move the knob, repaint is called.

https://puu.sh/yG6ZE/ca08ebd3f0.gif

omgaaad!! all I need is that repaint notification.

**Comment by RobinSchmidt** [comment]
the "modulatedValue" is actually updated per-sample - you certainly don't want to get per sample-notifications and call repaint per sample. do you want an animated gui that always shows the modulated value along with the slider center value?

i think, for the mod-system, it could make sense to send out notifications, when the mod-depth changes such that a slider can somehow indicate that depth (as mod-range), but an animation for the current modulated value is a different story. i'd probably try something based on juce::Timer to regularly poll the current value and repaint ...i'd probably make a class RepaintManager or something that can set a repaint interval and where sliders can register themselves to get called at regular intervals (and then register when there are connections and deregister when there are none). the RepaintManager would call repaint regularly on all its repaintees (of class juce::Component)...as a first idea

btw. - you did not say anything to my last question here: 
https://github.com/RobinSchmidt/RS-MET/issues/135#issuecomment-350587473

**Comment by elanhickler** [comment]
Yeah, maybe I should stick to mod depth change painting, because animation is going to be expensive.

You're talking about the invoice? Yes, send the invoice.

**Comment by elanhickler** [comment]
Actually I really like the idea of animation, it is just more informative and easier to grasp.

**Comment by RobinSchmidt** [comment]
yes, it's certainly a nice to have. actually, it would be best to have both

**Comment by RobinSchmidt** [comment]
looking at the inheritance diagram of this:
https://juce.com/doc/classAnimatedAppComponent
it seems i'm on the right track with juce::Timer. but it's probably wateful to have each widget its own timer, so registering at a global timer-derived RepaintManager is probably better

**Comment by RobinSchmidt** [comment]
btw - i also used juce::Timer for triggering the repaint in my PhaseScopeDisplay

**Comment by RobinSchmidt** [comment]
btw. - i'm currently adapting echolab for chainer and experimenting with other gui-colors (black/dark plots and widgets). black plots look much better, i think:

![image](https://user-images.githubusercontent.com/1833099/34048117-cbb73fba-e1b2-11e7-8c52-4b75d96f5b3c.png)

**Comment by elanhickler** [comment]
Yeah looks good. Are you using standalone app? Do you know how to make it use windows default?

**Comment by RobinSchmidt** [comment]
no, its the plugin. but i think, for standalone-apps to have native look-and-feel, you should have a look at:
https://juce.com/doc/classTopLevelWindow#a749fbd5e688ed8c9af3d0d99b21e18c8

**Comment by elanhickler** [comment]
I know, I already did that. Didn't work, that's the issue. 

Just add standalone to your build process in your jucer, just a matter of clicking a check box. And let me know if you figure out how to make it use windows default.

**Comment by RobinSchmidt** [comment]
where is it? i can't find it in the projucer. only the various plugin formats:
![image](https://user-images.githubusercontent.com/1833099/34050919-e8a79ae8-e1bc-11e7-8b99-f1eadb06b110.png)

**Comment by elanhickler** [comment]
I guess you have to upgrade your jucer.
![image](https://user-images.githubusercontent.com/9116575/34051021-434e16b8-e17a-11e7-869f-593abdc21b5d.png)

**Comment by elanhickler** [comment]
The only issue I had with upgaded jucer VS projects is it complaining that ProjectInfo is redefined, no idea how it manages to get redfined.... so I just rename to ProjectInfo2, nothing is affected by this rename....why.................

![image](https://user-images.githubusercontent.com/9116575/34051105-8d673b30-e17a-11e7-8641-247bee0e5768.png)

**Comment by RobinSchmidt** [comment]
wouldn't i have to upgrade my juce-version then too?

**Comment by elanhickler** [comment]
NOPE

**Comment by elanhickler** [comment]
I mean, upgrade your projucer, but leave your juce library code as is.

**Comment by elanhickler** [comment]
I think I can just send you the Projucer.exe so you don't have to create a juce repository and go througha ll those steps: http://www.elanhickler.com/transfer/Soundemote/Projucer.exe

**Comment by RobinSchmidt** [comment]
uuuhh...yes - the new version requires a login everytime you start it. annoying. do you use the self-compiled version with this gpl-flag? ...or what kind of licensing option do you use for juce?

**Comment by elanhickler** [comment]
i pay for a license... it is weird that it asks for login every time, it should save your info. I guess it doesn't?

**Comment by RobinSchmidt** [comment]
no - it doesn't save the login info here. i guess, if i start using it on a regular basis, i'll compile it for myself

**Comment by elanhickler** [comment]
o my god.

animated modulation indication is AWESOME!!!!!!!

It shows you exactly what's going on in the case of chaotic feedback routing especially, this is going to be so informative. I had no idea what kind of modulation curves were happening. Audio-rate feedback modulation can result in slow value changes!! This is eye opening!

**Comment by RobinSchmidt** [comment]
are you already doing it using my idea with the RapaintManager?

**Comment by elanhickler** [comment]
noooooooo, i;m just observing modulation by frantically moving the knob so it updates continuously. I think I'll have to wait for you to implement your idea, I don't know how to do it myself.

**Comment by RobinSchmidt** [comment]
i guess, making a standalone "plugin" have a native look requires to edit the juce library classes. i guess here:
```cpp
class StandaloneFilterApp  : public JUCEApplication
{
public:
    void initialise (const String&) override
    {
        mainWindow->setUsingNativeTitleBar (true);  
    }
```
...ohh...yes - it works

**Comment by RobinSchmidt** [comment]
however - i have no idea, how to do it without editing juce classes. maybe we can from the editor component somehow get a pointer to the top-level window and call setUsingNativeTitleBar on the returned pointer

**Comment by RobinSchmidt** [comment]
...ohhh - and the "Options" button is lost. not good. it's needed

**Comment by RobinSchmidt** [comment]
here:
https://juce.com/doc/classComponent#a64415a8c5911f94475b5c624598adfa8
then cast that pointer to TopLevelWindow and if the dynamic_cast returns a non-nullptr, call setUsingNativeTitleBar ...probably

**Comment by RobinSchmidt** [comment]
...but as said - without the options button, it's probably not a good idea to use the native window

**Comment by RobinSchmidt** [comment]
> we should try this again once we upgrade juce library.

indeed. it seems to make more sense to me to use the same version of jucer as the library. when i want to keep up with my plan to release chainer to christmas, i'll probably have to upgrade before and then use the "with-splashscreen" licensing option...or otherwise license my library under gpl or something (i actually want to use a more liberal, non-viral open-source license along with commercial licensing - we'll see).

**Comment by RobinSchmidt** [comment]
> i actually want to use a more liberal, non-viral open-source license along with commercial licensing

what i want, if someone uses my library for free in an open-source product, is that all the other code is also open-source, but not necessarily all gpl

**Comment by elanhickler** [comment]
Are you planning on having draggable components so we can reorder effects in plug.

also... not good:

![image](https://user-images.githubusercontent.com/9116575/34072645-4a832e2c-e248-11e7-9bb4-3d57af3ca69c.png)

**Comment by RobinSchmidt** [comment]
you mean, like in chainer, you could grab a slot and drag it to another slot to change slot-order? yeaahhh...that would be really nice to have. not trivial to implement, though - but eventually i really want that feature

**Comment by RobinSchmidt** [comment]
what do you mean by "not good"?

**Comment by elanhickler** [comment]
I need the previous color scheme on my plugins

**Comment by elanhickler** [comment]
I'm going to start replacing all your paint methods soon, so you won't have to worry about this in the future.

**Comment by RobinSchmidt** [comment]
really? you like the white plot-backgrounds more? there is some function to change the color-scheme, i'll have to dig it out..

btw - i wonder, if i should change my module-headline font, too. the big version of my pixel-font looks rather ugly (the small versions for the widgets and axis annotations are actually ok, imho)

**Comment by elanhickler** [comment]
The breakpoint window can be black, sure, but not all the sliders.

**Comment by RobinSchmidt** [comment]
> I'm going to start replacing all your paint methods soon, so you won't have to worry about this in the future.

for this to work with library-provided sub-editors (such as my breakpoint editor), i really need to delegate the painting to a "strategy" object (like we did with the parameter mapper) because you can't change the slider-class that is used in my breakpoint-editor

**Comment by RobinSchmidt** [comment]
because - you can, of course, override "paint" in a subclass derived from my slider class, say. but you can't tell the breakpoint editor to use your subclass for its sliders. that's why we need the strategy pattern for gui customization

**Comment by RobinSchmidt** [comment]
or i could templatize my editors on the widget classes to use. haha! nope! that's a fucking bad idea!

**Comment by elanhickler** [comment]
I don't need to tell breakpoint editor to use my subclass for its sliders. I would redesign the breakpoint editor from scratch.

**Comment by RobinSchmidt** [comment]
oh - ok - but that's a rather big task. i actually want to refactor all my plot-node-dragging editors at some point...factor out some common NodeEditor class to capture and re-use what they have in common. maybe with callbacks like "nodeDraggedTo(int index, int newX, int newY)". breakpoint-editor, equalizer-editor, etc could then be subclasses of that and just override these nodeDragged callbacks. all the mouse-event handling would be factored out, avoiding a lot of code-duplication

**Comment by elanhickler** [comment]
why would that be a big task? I don;t know when I'd tackle the task of creating my own breakpoint widget, but I can certainly create a new editor, put the breakpoint widget, and then put in whatever sliders I want around the breakpoint widget. Easy.

**Comment by RobinSchmidt** [comment]
well, i mean the breakpoint-widget itself is a quite big and complicated class - not putting some sliders around it

**Comment by elanhickler** [comment]
i never said I wanted to re-design the actual breakpoint widget, just the breakpoint module editor in the near future.

**Comment by RobinSchmidt** [comment]
ah - ok - misunderstanding

**Comment by RobinSchmidt** [comment]
call `setWidgetAppearance(ColourScheme::DARK_ON_BRIGHT);` somewhere in the constructor of your editor. ... but for some reason, the preset loader doesn't respect this setting. must be a bug - i will look into it...

**Comment by RobinSchmidt** [comment]
i'm not yet sure, if i'll keep using black widgets by default, though

**Comment by RobinSchmidt** [comment]
> for some reason, the preset loader doesn't respect this [colorscheme] setting. must be a bug - i will look into it...

fixed. also, in ToolChain, you can now interactively edit the colorscheme settings (using the "Setup" button)

**Comment by RobinSchmidt** [comment]
...i think, tomorrow, i'll attempt the upgrade to juce 5. if that works smoothly, the next step is to get all the preset- and sample-directory stuff for mac right

**Comment by elanhickler** [comment]
yayyyyyyyyyy!!!!

**Comment by elanhickler** [comment]
did you upgrade yet? should i pull latest from repository?

**Comment by RobinSchmidt** [comment]
yes, i updated the juce version on tuesday

**Comment by elanhickler** [comment]
can't build, some weird errors form juce library. How did you get your project to build? 

Severity	Code	Description	Project	Line	File	Suppression State
Error	C3805	'type': unexpected token, expected either '}' or a ',' (compiling source file ..\..\JuceLibraryCode\include_se_framework.cpp)	Spiral Generator_SharedCode	288	d:\_programming\rs-met\libraries\juce\modules\juce_gui_extra\misc\juce_PushNotifications.h	

Error	C2065	'large': undeclared identifier (compiling source file ..\..\JuceLibraryCode\include_se_framework.cpp)	Spiral Generator_SharedCode	292	d:\_programming\rs-met\libraries\juce\modules\juce_gui_extra\misc\juce_PushNotifications.h	

Error	C2062	type 'char' unexpected (compiling source file ..\..\JuceLibraryCode\include_se_framework.cpp)	Spiral Generator_SharedCode	288	d:\_programming\rs-met\libraries\juce\modules\juce_gui_extra\misc\juce_PushNotifications.h

**Comment by RobinSchmidt** [comment]
i never hit this with ToolChain. no idea why. i have just tried to build SpiralGenerator and hit this error there, too. it seems in juce_PushNotifications.h there's an enum (line 285):

```
        enum BadgeIconType
        {
            none,
            small,
            large
        };
```

and there seems to be a macro-definition for `small` in rpcnr.h (no idea where this file is included or is located or what it is good for). hmmm...

**Comment by RobinSchmidt** [comment]
so, i guess this macro-definition causes the keyword `char` to appear in an enum (after the preprocessor stage has passed) and that causes an error. i don't know, why i don't hit this

**Comment by elanhickler** [comment]
What if it's that I need to create my jucer project from scratch? The problem is that my shared framework stops working, bajillion errors.

**Comment by RobinSchmidt** [comment]
> What if it's that I need to create my jucer project from scratch?

why should you have to do that? i didn't do that either.

**Comment by RobinSchmidt** [comment]
actually, i get only 6 errors - but zillions of warnings

**Comment by RobinSchmidt** [comment]
...nope - that's not it. i removed those modules, the error remains. ...but maybe it's a good idea to remove those modules anyway, when they are not needed for that project.

**Comment by elanhickler** [comment]
indeed 👍

**Comment by RobinSchmidt** [comment]
oh - i just see juce_audio_utils is actually needed for the standalone build

**Comment by RobinSchmidt** [comment]
ok, pull an update of my library and try again. i hacked juce (added an #undef small) which seems to fix the *weird* error. you'll still get some other errors but these are apparently related to some setup-stuff in your jucer file. you can remove the crypto and video modules, too.

**Comment by elanhickler** [comment]
oooooooooooooook what about prettyscope?

**Comment by RobinSchmidt** [comment]
well, there, you need opengl, of course!

**Comment by elanhickler** [comment]
can you also fix this white text on white background, agghhh!!!

![image](https://user-images.githubusercontent.com/9116575/34282118-725bf12c-e680-11e7-8aed-04461c2c99b7.png)

**Comment by RobinSchmidt** [comment]
iiihh! is this new?

**Comment by elanhickler** [comment]
of course, it's when you started trying dark theme for your plugins

**Comment by elanhickler** [comment]
Robin, do you have in place modulation repainting methods for modulation indicators that I can override?

**Comment by RobinSchmidt** [comment]
do you mean the color overlay when a slider has modulation connections (because that other kind of modulation indication does not exist (yet)). if so, then you can just override ModulatableSlider::paint

**Comment by elanhickler** [comment]
I mean animated modulation indication, realtime update.

![image](https://camo.githubusercontent.com/b5aff80d06c21a5b011037cdcb062f7bf42dd455/68747470733a2f2f7075752e73682f7947365a452f636130386562643366302e676966)

**Comment by elanhickler** [comment]
Are you planning on adding modulation callback for animation soon? 😃

**Comment by RobinSchmidt** [comment]
well, i wanted to look into spline-drawing for scope-oversampling next. what's your priority list?

**Comment by elanhickler** [comment]
Priority: Whatever is easier/faster. I can do an update to all my oscilloscope music plugins if you add modulation indicator to your default sliders. That will be a huge and appreciated addition.

**Comment by RobinSchmidt** [comment]
i think, scope interpolation is easier. it's just some straightforward math. ...ok - for a more refined version, maybe also some not-so-straightforward stuff - but a first version may not need that. the modulation animation is a software engineering/design problem and i'm not yet sure, what's the best way to approach it (but i have an idea)...but it could be fiddly

**Comment by elanhickler** [comment]
What are the main design problems? CPU usage? Class relationships?

**Comment by RobinSchmidt** [comment]
yes. both. probably also multithreading. i guess i'll introdcue some sort of repaint-manager class that regularly calls the repaint callback on some number of registered repaintees (the widgets)....but better call the repaint callbacks only when actually necessary (the (modulated) value actually changed)

**Comment by elanhickler** [comment]
Ok, work on what you are most inspired to work on.  Work on Liberty if you want to create a powerful prototyping environment. Work on the oscilloscope if you want a better visual. Work on modulation animation if you want to bring your framework to the next level.

**Comment by elanhickler** [comment]
I actually do really need modulation indication by November 1st. That gives you 7 weeks and 2 days!

**Comment by RobinSchmidt** [comment]
ok, let me finish the spline drawing for the scopes and then i'll look into this

**Comment by elanhickler** [comment]
question: can you make phase scope drawing a separate process as you look into modulation indication? It's kinda a similar thing, right? Having multiple phase scopes displaying in a project can start to stop the DAW UI from updatin.

**Comment by RobinSchmidt** [comment]
i guess, you mean a separate *thread*, not *process*, right? because i think it would be weird for an audio plugin to spawn processes. i just checked the threading model of my scope. all the real drawing work is actually done in the audio thread. in the gui-thread, i just draw the ready-made juce::Image to the screen which shouldn't be that expensive. ...unless the resampling is expensive (but that should apply only, when the "Scale" parameter is something other than 1 (in this case, i draw onto a lower-resolution image (making my drawing code less expensive) and let juce upsample it)

the modulation indication, as i currently have laid it out in my mind, would also draw in the gui thread. in my mixsonic mini-daw, i actually didi something like that - i delegated the drawing to another thread because indeed the gui became unresponsive when moving around the clips (each of which has to draw a waveform display). i guess i could dig out that code. ...i was a bit messy but sorta worked. i actually want to integrate that code into my codebase anyway

**Comment by elanhickler** [comment]
So how is your idea different that just... say... every 1/30 of a second, get all modulated parameter settings, call repaint?

**Comment by RobinSchmidt** [comment]
well....yeah...you can inherit your widget from juce::Timer and repaint every 1/30 seconds.....but i think that would be wasteful if there's nothing to actually redraw. i think the system should figure out, if redrawing is needed and redraw only then

**Comment by RobinSchmidt** [comment]
i think, that would be a very dirty solution. it may work, though

**Comment by elanhickler** [comment]
So then a simple if statement to start if connections exist, stop timer if no connections exist.

**Comment by RobinSchmidt** [comment]
i'm currently setting up a RepaintManager class. i think, it is better to have one central object that derives from Timer and that calls repaint on a number of registered "repaintee" objects - so we don't need to bloat each and every of our widget classes by adding yet another baseclass. probably much leaner design, memory footprint wise and also probably less code to write

**Comment by RobinSchmidt** [comment]
...although...looking at the Timer class - it actually looks quite lean....maybe my design idea is needlessly complicated...hmmm

**Comment by elanhickler** [comment]
hmmmmm.... I'm thinking either 30fps or sync timer to monitor refresh rate (using that bool switch in juce)

**Comment by elanhickler** [comment]
I see you added a modulatable slider animated class, which inherits timer. Why? Don't we want a global timer that starts if `modulation connections > 0` and stops if not? Then upon timer trigger, repaint.

**Comment by RobinSchmidt** [comment]
i don't know...i was starting to implement your design idea....do you think, i should instead follow my (more complicated) idea instead?

**Comment by RobinSchmidt** [comment]
do you think, i should follow the ideas, i outlined with the RepaintManager? (i actually already created a stub for such a class)....yeahhh....maybe that's a better idea. maybe some other client code already want to use the Time baseclass for something else and it just seems to be an overall cleaner design

**Comment by elanhickler** [comment]
Yes. I need to repaint constantly more than just modulatable sliders, repaint manager could work.

**Comment by RobinSchmidt** [comment]
yes - i'll go that route with the repaint manager. has also the advantage that we can just de/register at the manager whenever the mod-connection status changes. the widget itself doesn't need to keep track of that and we'll iterate only through those widgets for repainting that actually need repaint
